compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp

```diff
@@ -205,54 +189,132 @@ LogicalResult fuseForallIntoSlice(RewriterBase &rewriter,
+  //   %p:n = affine.delinearize_index %id into [%P0, ..., %Pn]
+  //   ...
+  // }
+  //
+  // Note: We use 0 as the loop lower bound instead of the linearized consumer
+  // loop ID if possible to make later loop promotion patterns easier.
 
   MLIRContext *context = rewriter.getContext();
   Location loc = producer.getLoc();
 
+  // Compute the linearize consumer loop ID and total consumer loop worker
+  // count (C0 * ... * Cn).
   AffineExpr d0, d1, d2;
   bindDims(context, d0, d1, d2);
   AffineExpr mulAdd = d0 * d1 + d2;
   OpFoldResult linearId = rewriter.getIndexAttr(0);
+  OpFoldResult consumerWorkerCount = rewriter.getIndexAttr(1);
   for (auto loop : *consumerLoopNest) {
     for (auto [inductionVar, workerCount] :
          llvm::zip_equal(getAsOpFoldResult(loop.getInductionVars()),
                          loop.getMixedUpperBound())) {
       linearId = affine::makeComposedFoldedAffineApply(
           rewriter, loc, mulAdd, {linearId, workerCount, inductionVar});
+      consumerWorkerCount = affine::makeComposedFoldedAffineApply(
+          rewriter, loc, d0 * d1, {consumerWorkerCount, workerCount});
     }
   }
 
-  Value linearThreadIdVal =
+  // Compute the total producer loop worker count (P0 * ... * Pn).
+  Value linearConsumerIdVal =
       getValueOrCreateConstantIndexOp(rewriter, loc, linearId);
-  SmallVector<Value> ranges;
-  for (auto workerCount : producer.getStaticUpperBound()) {
-    ranges.push_back(rewriter.create<arith::ConstantIndexOp>(loc, workerCount));
+  SmallVector<Value> producerRanges;
+  OpFoldResult producerWorkerCount = rewriter.getIndexAttr(1);
+  for (auto workerCount : producer.getMixedUpperBound()) {
+    producerRanges.push_back(
+        getValueOrCreateConstantIndexOp(rewriter, loc, workerCount));
+    producerWorkerCount = affine::makeComposedFoldedAffineApply(
+        rewriter, loc, d0 * d1, {producerWorkerCount, workerCount});
   }
-  ValueRange newIds = rewriter
-                          .create<affine::AffineDelinearizeIndexOp>(
-                              loc, linearThreadIdVal, ranges)
-                          .getResults();
-
-  // Step 2. Inline the region of the producer.
-  SmallVector<Value> bbArgReplacements(newIds);
-  bbArgReplacements.append(producer.getOutputs().begin(),
-                           producer.getOutputs().end());
 
+  std::optional<int64_t> staticProducerCount =
+      getConstantIntValue(producerWorkerCount);
+  std::optional<int64_t> staticConsumerCount =
+      getConstantIntValue(consumerWorkerCount);
+  bool perfectlyDivides =
+      staticConsumerCount && staticProducerCount &&
+      staticProducerCount.value() % staticConsumerCount.value() == 0;
+
+  // Step 3. Create the `scf.for` loop for the producer.
+  // If the consumer worker count perfectly divides the producer worker count,
+  // then we can use a lower bound of 0 and keep the loop bounds static.
+  Value lb = perfectlyDivides ? rewriter.create<arith::ConstantIndexOp>(loc, 0)
```

Comment:
I think this is a premature optimization. You should always use `linearConsumerIdVal` and then if you need to normalize the loop after the fact?

Reply:
The issue is that folding the loop without doing this is quite difficult without an `affine.linearize_index` op and a framework for inferring int ranges. We will have IR that looks something like this
```
%id = affine.linearize_index %0, %1, ... basis = [%C0, %C1, ...]
scf.for %i = %id to %C0 * %C1 * ... step %P0 * %P1 * ...
```
Lets take the simple case of `%C0 * %C1 * ... == %P0 * %P1 * ...` (same loop trip count). To fold the loop away, we would have to prove that `%id` is in the range `[0, %C0 * %C1 * ...)` which is very difficult without proper index range analysis. Conversely it's very easy to do that folding at this point when we're forming the loop.